Code/src/main/java/nl/utwente/group10/haskell/expr/Function.java

```diff
@@ -0,0 +1,167 @@
+package nl.utwente.group10.haskell.expr;
+
+import nl.utwente.group10.haskell.env.Env;
+import nl.utwente.group10.haskell.exceptions.HaskellException;
+import nl.utwente.group10.haskell.hindley.GenSet;
+import nl.utwente.group10.haskell.hindley.HindleyMilner;
+import nl.utwente.group10.haskell.type.FuncT;
+import nl.utwente.group10.haskell.type.Type;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Represents a user-defined Haskell function. Holds arguments which can be used for building the function.
+ */
+public class Function extends Expr {
+
+    /**
+     * An argument for a function. Can be used as a variable in a function definition.
+     */
+    public class FunctionArgument extends Expr {
+
+        /** The function this argument is for. */
+        private Function function;
+
+        /**
+         * @param function The function this argument is for.
+         */
+        protected FunctionArgument(Function function) {
+            this.function = function;
+        }
+
+        /**
+         * @return The 0-indexed position of this argument in the function definition.
+         */
+        public int getPosition() {
+            return this.function.arguments.indexOf(this);
+        }
+
+        @Override
+        public Type analyze(Env env, GenSet genSet) throws HaskellException {
+            return HindleyMilner.makeVariable(); // TODO Make sure that this is correctly unified and/or make this editable.
+        }
+
+        @Override
+        public String toHaskell() {
+            return String.format("%s%d", Function.ARG_PREFIX, this.getPosition());
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s%d", Function.ARG_PREFIX, this.getPosition());
+        }
+    }
+
+    /** The prefix for function argument names. */
+    protected static final String ARG_PREFIX = "arg";
+
+    /** The arguments for this function. */
+    private List<FunctionArgument> arguments;
+
+    /** The expression that forms the base of this function. */
+    private Optional<Expr> expr;
+
+    /**
+     * @param expr The expression that forms the base of this function.
+     */
+    public Function(Expr expr) {
+        this.arguments = new ArrayList<>();
+        this.expr = Optional.ofNullable(expr);
+    }
+
+    /**
+     */
+    public Function() {
+        this(null);
+    }
```

Comment:
Hmm, yes, very true. 

Here's an idea. I don't like it very much myself, but maybe it sparks some inspiration. You could allow frontend to build FunctionArgument exprs all by itself, making FunctionArgument basically nothing more than a wrapped integer *n*. A FunctionArgument that wraps a 1 would mean 'the second argument to a function, whatever it's going to be'. You'd collect these things at `Function` construction time, count them (the function will have `max(n)+1` args), attach some types, do some inference and so on and so forth.

Practically this would put the egg before the chicken. Does that make sense?